### ISSUE TITLE
Astaire crashes under load

### ISSUE THREAD

**Issue by rkd-msw** [Metaswitch/astaire#71]
#### Symptoms

@eleanor-merry was running Astaire under load (and can probably give more details). Astaire crashed with the following backtrace:

```
#0  0x00007fd065becc37 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007fd065bf0028 in __GI_abort () at abort.c:89
#2  0x000000000042842a in signal_handler (sig=11) at main.cpp:212
#3  <signal handler called>
#4  0x000000000042b4f3 in _M_lower_bound (this=<optimized out>, __k=<error reading variable: Cannot access memory at address 0x0>, __y=0x7fd0206969c0, __x=0x7fd0201c9910)
    at /usr/include/c++/4.8/bits/stl_tree.h:1140
#5  lower_bound (__k=<error reading variable: Cannot access memory at address 0x0>, this=0x7fd0206969b8) at /usr/include/c++/4.8/bits/stl_tree.h:879
#6  lower_bound (__x=<error reading variable: Cannot access memory at address 0x0>, this=0x7fd0206969b8) at /usr/include/c++/4.8/bits/stl_map.h:864
#7  operator[] (__k=<error reading variable: Cannot access memory at address 0x0>, this=0x7fd0206969b8) at /usr/include/c++/4.8/bits/stl_map.h:461
#8  MemcachedBackend::get_replicas (this=this@entry=0x22b5190, vbucket=vbucket@entry=26, operation=operation@entry=MemcachedBackend::READ) at memcached_backend.cpp:284
#9  0x000000000042bea8 in MemcachedBackend::read_data (this=0x22b5190, key="reg\\\\sip:2010000278@em7.cw-ngv.com", data="", cas=@0x7fbdfe3c7e28: 1) at memcached_backend.cpp:383
#10 0x00000000004290dc in ProxyServer::handle_get (this=this@entry=0x22b5130, get_req=0x7fd02020aca0, connection=connection@entry=0x7fd038036d90) at proxy_server.cpp:330
#11 0x0000000000429833 in ProxyServer::connection_thread_fn (this=0x22b5130, connection=0x7fd038036d90) at proxy_server.cpp:253
#12 0x0000000000429a90 in ProxyServer::connection_thread_entry_point (params_arg=0x7fd03801d4b0) at proxy_server.cpp:224
#13 0x00007fd066d17184 in start_thread (arg=0x7fbdfe3c8700) at pthread_create.c:312
#14 0x00007fd065cb037d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

I jumped to frame 8 and checked `_read_replicas`, and it looks corrupt:

```
(gdb) p _read_replicas
$7 = std::vector of length 128, capacity 128 = {std::vector of length 0, capacity -17566630078711, std::vector of length 4550841, capacity 4550841 = {Cannot access memory at address 0x0
```

#### Impact

Astaire crashes. The biggest impact of that is likely to be a CPU% increase as we gather diags.

#### Release and environment 

@eleanor-merry ?

#### Steps to reproduce

@eleanor-merry ?

**Comment by rkd-msw** [comment]
I wonder if this might be some kind of threading issue - sensible next step would be to run Astaire under helgrind, maybe?

**Comment by AlexHockey** [comment]
I've only had a chance to do some code reading and thinking, but I haven't made much progress. 

From gdb it appears that the `_read_replicas` is ok (it has the right length and capacity) but the contents have become corrupted. `_read_replicas` is fixed size after it is constructed, so I think we can rule out the possibility that it's being resized on one thread and accessed on another. So I think we've either put some garbage into the vector, or there's a data race between a write and a read. Writes only only happen [here](https://github.com/Metaswitch/astaire/blob/b8723f96ee3ec350a61361f3d2db6b6fdae05479/src/memcached_backend.cpp#L152) but this code is protected by a lock (specifically a rw lock that is held by locked for a writer), and the only reads from the vector are done when the lock is held for a reader. This implies that garbage has been put into the vector. The vector is populated from the contents of a `MemcachedStoreView`, but this object is only executed on a single thread. 

TL;DR - I have no idea what's going on. Memcheck and helgrind (or someone smarter than me) are needed.

**Comment by AlexHockey** [comment]
I've run Astaire under memcheck and helgrind. Memcheck showed some leaks but nothing that ought to cause a crash. Helgrind showed a load of data races (most of which were in zmq) but I'll look thought these tomorrow. 

[astaire-helgrind.txt](https://github.com/Metaswitch/astaire/files/590107/astaire-helgrind.txt)
[astaire-memcheck.txt](https://github.com/Metaswitch/astaire/files/590108/astaire-memcheck.txt)

**Comment by AlexHockey** [comment]
The helgrind output contains:

* A bunch of data races that are deep in zmq. I've seen these before in other components (such as sprout) and never known them to cause a problem. 
* Some data races in `AstaireGlobalStatistics`. These all refer to stores/loads of atomic variables. I'm not sure why helgind is picking these up, but it seems unlikely these are the culprit.
* This data race. This is caused by unsynchronized access to `MemcachedBackend::_view_number`. This is not an atomic variable. I don't see how this not being atomic could cause problems - a torn read or a missed increment would just cause an update to be ignored.  

```
==11032== Possible data race during write of size 8 at 0x675FD50 by thread #15
==11032== Locks held: 1, at address 0x675FD58
==11032==    at 0x42AE50: MemcachedBackend::new_view(MemcachedConfig const&) (memcached_backend.cpp:159)
==11032==    by 0x42AF8C: MemcachedBackend::update_config() (memcached_backend.cpp:171)
==11032==    by 0x42D159: operator() (stl_function.h:551)
==11032==    by 0x42D159: updater (updater.h:103)
==11032==    by 0x42D159: Updater<void, MemcachedBackend>::updater_thread(void*) (updater.h:85)
==11032==    by 0x4C30FA6: ??? (in /usr/lib/valgrind/vgpreload_helgrind-amd64-linux.so)
==11032==    by 0x5086183: start_thread (pthread_create.c:312)
==11032==    by 0x5F1437C: clone (clone.S:111)
==11032==
==11032== This conflicts with a previous read of size 8 by thread #20
==11032== Locks held: none
==11032==    at 0x42B45A: MemcachedBackend::get_replicas(int, MemcachedBackend::Op) (memcached_backend.cpp:214)
==11032==    by 0x42C4C7: MemcachedBackend::read_data(std::string const&, std::string&, unsigned long&) (memcached_backend.cpp:383)
==11032==    by 0x4296FB: ProxyServer::handle_get(Memcached::GetReq*, Memcached::ServerConnection*) (proxy_server.cpp:330)
==11032==    by 0x429E52: ProxyServer::connection_thread_fn(Memcached::ServerConnection*) (proxy_server.cpp:253)
==11032==    by 0x42A0AF: ProxyServer::connection_thread_entry_point(void*) (proxy_server.cpp:224)
==11032==    by 0x4C30FA6: ??? (in /usr/lib/valgrind/vgpreload_helgrind-amd64-linux.so)
==11032==    by 0x5086183: start_thread (pthread_create.c:312)
==11032==    by 0x5F1437C: clone (clone.S:111)
==11032==  Address 0x675fd50 is 48 bytes inside a block of size 288 alloc'd
==11032==    at 0x4C2C500: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_helgrind-amd64-linux.so)
==11032==    by 0x407E6B: main (main.cpp:332)
==11032==  Block was alloc'd by thread #1
==11032==
```
* This data race (and another similar one). I don't understand this one at all and it needs more investigation. 
```
==11032== Possible data race during write of size 8 at 0x67636F0 by thread #7
==11032== Locks held: none
==11032==    at 0x56C135F: ??? (in /usr/lib/x86_64-linux-gnu/libzmq.so.3.1.0)
==11032==    by 0x56BE90D: ??? (in /usr/lib/x86_64-linux-gnu/libzmq.so.3.1.0)
==11032==    by 0x56BDBAD: ??? (in /usr/lib/x86_64-linux-gnu/libzmq.so.3.1.0)
==11032==    by 0x56E21DF: ??? (in /usr/lib/x86_64-linux-gnu/libzmq.so.3.1.0)
==11032==    by 0x4C30FA6: ??? (in /usr/lib/valgrind/vgpreload_helgrind-amd64-linux.so)
==11032==    by 0x5086183: start_thread (pthread_create.c:312)
==11032==    by 0x5F1437C: clone (clone.S:111)
==11032==
==11032== This conflicts with a previous read of size 8 by thread #1
==11032== Locks held: none
==11032==    at 0x40E564: ~vector (stl_vector.h:416)
==11032==    by 0x40E564: _Destroy<std::vector<std::basic_string<char> > > (stl_construct.h:93)
==11032==    by 0x40E564: __destroy<std::vector<std::basic_string<char> >*> (stl_construct.h:103)
==11032==    by 0x40E564: _Destroy<std::vector<std::basic_string<char> >*> (stl_construct.h:126)
==11032==    by 0x40E564: _Destroy<std::vector<std::basic_string<char> >*, std::vector<std::basic_string<char> > > (stl_construct.h:151)
==11032==    by 0x40E564: ~vector (stl_vector.h:415)
==11032==    by 0x40E564: MemcachedStoreView::~MemcachedStoreView() (memcachedstoreview.cpp:63)
==11032==    by 0x42AE64: MemcachedBackend::new_view(MemcachedConfig const&) (memcached_backend.cpp:161)
==11032==    by 0x42AF8C: MemcachedBackend::update_config() (memcached_backend.cpp:171)
==11032==    by 0x42AB4E: operator() (stl_function.h:551)
==11032==    by 0x42AB4E: Updater (updater.h:61)
==11032==    by 0x42AB4E: MemcachedBackend::MemcachedBackend(MemcachedConfigReader*, BaseCommunicationMonitor*, Alarm*) (memcached_backend.cpp:94)
==11032==    by 0x407E83: main (main.cpp:332)
==11032==  Address 0x67636f0 is 384 bytes inside a block of size 400 alloc'd
==11032==    at 0x4C2BFA0: malloc (in /usr/lib/valgrind/vgpreload_helgrind-amd64-linux.so)
==11032==    by 0x56C113C: ??? (in /usr/lib/x86_64-linux-gnu/libzmq.so.3.1.0)
==11032==    by 0x56C635E: ??? (in /usr/lib/x86_64-linux-gnu/libzmq.so.3.1.0)
==11032==    by 0x56C0DDF: ??? (in /usr/lib/x86_64-linux-gnu/libzmq.so.3.1.0)
==11032==    by 0x56D289B: ??? (in /usr/lib/x86_64-linux-gnu/libzmq.so.3.1.0)
==11032==    by 0x56D5E68: ??? (in /usr/lib/x86_64-linux-gnu/libzmq.so.3.1.0)
==11032==    by 0x56E98AB: ??? (in /usr/lib/x86_64-linux-gnu/libzmq.so.3.1.0)
==11032==    by 0x56E9974: zmq_send (in /usr/lib/x86_64-linux-gnu/libzmq.so.3.1.0)
==11032==    by 0x426463: AlarmReqAgent::agent() (alarm.cpp:421)
==11032==    by 0x426828: AlarmReqAgent::agent_thread(void*) (alarm.cpp:318)
==11032==    by 0x4C30FA6: ??? (in /usr/lib/valgrind/vgpreload_helgrind-amd64-linux.so)
==11032==    by 0x5086183: start_thread (pthread_create.c:312)
==11032==  Block was alloc'd by thread #5
```

**Comment by AlexHockey** [comment]
I can't figure out why that last data race is happening (maybe it is a red-herring).

**Comment by AlexHockey** [comment]
I've gone back to the original core file. The first thing I noticed is that *two* threads (both connection handling threads) are in the middle of handling a SIGSEGV. I also spotted that the main thread has the following stack:

```
Thread 6 (Thread 0x7fd067380780 (LWP 16863)):
#0  0x00007fd066d1865b in pthread_join (threadid=140532961838848, thread_return=0x0) at pthread_join.c:92
#1  0x0000000000418f52 in LastValueCache::~LastValueCache (this=0x22ae420, __in_chrg=<optimized out>) at ../modules/cpp-common/src/zmq_lvc.cpp:103
#2  0x0000000000407f2f in main (argc=<optimized out>, argv=<optimized out>) at main.cpp:362
```

This shows that it has gotten past the termination semaphore, which suggests its been sent a SIGTERM. It's then gone on to delete the `ProxyServer` (that the connection threads are executing code in) and everything blows up. 

It looks like it was queue manager that triggered the restart:

```
18-10-2016 16:56:02.885 UTC INFO apply_config_plugin.py:52 (thread ApplyConfigPlugin): Restarting services
```

This coincides with the timestamp on the corefile

```
-rw------- 1 ubuntu ubuntu  98M Oct 18 16:56 core.astaire.1476809776.gz
```

We already knew that Astaire does not exit cleanly, and there's a 2pt story in the backlog to fix this up.

**Comment by AlexHockey** [comment]
So normally we don't care too much about crashes during termination. This one was a bit interesting because it ended up with the core file being 73GB in size when unzipped. My guess is that when the `ProxyServer` got deleted we scribbled and accidentally allocated a shed-load of memory. But that could happen in any crash. I'm also not convinced it caused a 73GB file on disk - this would have been more than the total size of the disk (most of the file is zeros which I'm guessing the disk can optimize away). 

TL;DR I think this should be downgraded, but it might be worth scheduling the story to make astaire terminate cleanly.

**Comment by rkd-msw** [comment]
Removing the critical label as this only affects shutdown.
